Dockerize the app

achirazi opened the pull request.

smoqadam reviewed (commented):
Since we renamed the `opensearch-node` to `opensearch` in docker-compose.yaml, we must rename it everywhere else such as other services in docker-compose.yaml and config.php.dist, etc.

achirazi commented:
> Since we renamed the `opensearch-node` to `opensearch` in docker-compose.yaml, we must rename it everywhere else such as other services in docker-compose.yaml and config.php.dist, etc.

Fixed it. I went through all the files. If I missed something, please add another comment to that specific line.

Are you happy with the README.md that I've added, or do I need to add more info?

nlisgo commented on composer.json:
```diff
@@ -46,12 +46,13 @@
     "config": {
         "sort-packages": true,
         "allow-plugins": {
-            "mindplay/composer-locator": true
```
Reinstate this.

achirazi replied:
Fixed in last commit

nlisgo commented on composer.json:
```diff
@@ -46,12 +46,13 @@
         "symfony/browser-kit": "^3.1",
-        "symfony/var-dumper": "^3.1"
+        "symfony/var-dumper": "^3.1",
+        "ext-gearman": "*"
```
Would expect this to introduce a change to composer.lock

nlisgo commented on README.md:
```diff
@@ -0,0 +1,17 @@
+eLife Search
+=============
+
+This project is using elasticsearch to index content of the eLife articles and provide a full-text search for Journal.
```
It's important that we state that docker is just used to improve the developer experience and that to reliably recreate any issue experienced in ci or prod we should continue to use builder.

achirazi replied:
Added that in my last commit

nlisgo commented on README.md:
```diff
@@ -0,0 +1,17 @@
+1. Clone the project `git clone https://github.com/elifesciences/search.git`
+2. Run `docker-compose up --build`
+
+Important : Rename `config.php.dist` on local to `config.php` because config.php is in .gitignore.
```
This step should probably go before the docker command.

nlisgo replied:
How should we access the search api?

http://localhost:????

achirazi replied:
Added step before docker command.

The search API from what @smoqadam shared with me is not accessible from browser.
We had to remove gearman-ui and opensearch-dashboard due to the fact that if we disable SSL for opensearch, the dashboard also needs to be configured to connect to opensearch without ssl (I don't know if that's possible).

nlisgo commented on src/Search/Kernel.php:
```diff
@@ -399,10 +400,10 @@ public function dependencies(Application $app)
-                    'key' => $app['config']['aws']['key'],
-                    'secret' => $app['config']['aws']['secret'],
-                ];
+                $config['credentials'] = new Credentials(
```
If not sure why this is necessary. Will investigate when I can access the app.

achirazi replied:
Reverted changes in last commit. @smoqadam told me that he will include this change in his current branch

nlisgo commented on Dockerfile:
```diff
@@ -0,0 +1,11 @@
+
+COPY .docker/php.ini /usr/local/etc/php/php.ini
+
+WORKDIR /app
```
Since the Dockerfile and docker-compose.yaml are just needed for developer experience I think we should consider placing them in a subfolder or renaming them so they are not automatically picked up. We can still advise in the README how the docker container can be started. Let's discuss tomorrow.

achirazi replied:
Is there any other way to set up docker for development? 

I don't have enough info on search. I'm happy to insert any additional info inside Readme, just feel free to let me know what needs to be added.

achirazi replied:
Created the subfolder and added the file in there in my last commit

smoqadam commented on README.md:
```diff
@@ -0,0 +1,20 @@
+
+1. Clone the project `git clone https://github.com/elifesciences/search.git`
+2. Rename `config.php.dist` on local to `config.php` because config.php is in .gitignore.
+3. Run `docker-compose -f ./dev/docker-compose.yml up --build`
```
it's `docker-compose.yaml`

nlisgo commented on dev/docker-compose.yaml:
```diff
@@ -0,0 +1,75 @@
+            dockerfile: Dockerfile
+        volumes:
+            - ..:/app
+            - ./config.php:/app/config.php
```
I haven't confirmed that this works yet. I am on a train at the moment so I will try this out when I'm not on the move.

nlisgo replied:
This seems to be working as expected.

smoqadam replied:
What is the reason for putting `config.php.dist` into `dev` folder? If someone wants to run the app with `builder`, they need to copy it into the root folder. Anyhow, if we decide to keep it in `dev`, then, we should change this line to `- ./config.php.dist:/app/config.php`

nlisgo commented on dev/.docker/goaws.yaml:
```diff
@@ -0,0 +1,10 @@
+  Host: goaws
+  Port: 4100
+  # SqsPort: 9324
+  # SnsPort: 9292
```
Why include these if we are not using them?

achirazi replied:
Fixed in last commit

nlisgo commented on dev/.docker/goaws.yaml:
```diff
@@ -0,0 +1,10 @@
+Dev:                                # Another environment
```
Get rid of `# Another environment` comment

achirazi replied:
Fixed in last commit

nlisgo commented:
I'm hitting this error:

```
Error response from daemon: Conflict. The container name "/goaws" is already in use by container "7ad115f130e3e42ab7609f50260337225eab0df9e9929e1691fc6a43315244d2". You have to remove (or rename) that container to be able to reuse that name.
```

On running: `docker compose -f dev/docker-compose.yaml up --build`

Is this the issue you are experiencing @smoqadam ?

nlisgo commented on README.md:
````diff
@@ -0,0 +1,48 @@
+2. Run following commands in order:
+
+```bash 
+$ bin/console queue:create
````
Getting the following error at this step: 

```
root@e0bbc1a8fb5e:/app# bin/console queue:create

Fatal error: Uncaught Elasticsearch\Common\Exceptions\NoNodesAvailableException: No alive nodes found in your cluster in /app/vendor/elasticsearch/elasticsearch/src/Elasticsearch/ConnectionPool/StaticNoPingConnectionPool.php:64
Stack trace:
#0 /app/vendor/elasticsearch/elasticsearch/src/Elasticsearch/Transport.php(82): Elasticsearch\ConnectionPool\StaticNoPingConnectionPool->nextConnection()
#1 /app/vendor/elasticsearch/elasticsearch/src/Elasticsearch/Transport.php(99): Elasticsearch\Transport->getConnection()
#2 /app/vendor/elasticsearch/elasticsearch/src/Elasticsearch/Connections/Connection.php(297): Elasticsearch\Transport->performRequest('GET', '/key_value_stor...', Array, NULL, Array)
#3 /app/vendor/react/promise/src/FulfilledPromise.php(28): Elasticsearch\Connections\Connection->Elasticsearch\Connections\{closure}(Array)
#4 /app/vendor/ezimuel/ringphp/src/Future/CompletedFutureValue.php(55): React\Promise\FulfilledPromise->then(Object(Closure), NULL, NULL)
#5 /app/vendor/ezimuel/ringphp/src/Core.php(341): GuzzleHttp in /app/vendor/elasticsearch/elasticsearch/src/Elasticsearch/ConnectionPool/StaticNoPingConnectionPool.php on line 64
```

smoqadam replied:
@nlisgo It happens because it takes some time for elasticsearch node to be ready, if give it a minute or two, you won't see this error.

nlisgo commented:
> I'm hitting this error:
> 
> ```
> Error response from daemon: Conflict. The container name "/goaws" is already in use by container "7ad115f130e3e42ab7609f50260337225eab0df9e9929e1691fc6a43315244d2". You have to remove (or rename) that container to be able to reuse that name.
> ```
> 
> On running: `docker compose -f dev/docker-compose.yaml up --build`
> 
> Is this the issue you are experiencing @smoqadam ?

I removed the existing containers with the same name and no longer get this error.

nlisgo commented on README.md:
```diff
@@ -9,8 +9,9 @@ To reliably recreate any issue experienced in CI or Prod you should continue to
-2. Rename `/dev/config.php.dist` on local to `/dev/config.php` 
-3. Run `docker-compose -f dev/docker-compose.yaml up --build`
+2. Rename `/dev/config.php.dist` on local to `/dev/config.php`
+3. Run `mkdir logs && sudo chmod 777 ./logs`
```
We could add logs to dev folder with .gitkeep if needed? This would avoid the need to have this step but will try this out.

nlisgo commented on README.md:
```diff
@@ -0,0 +1,49 @@
+4. Run `docker-compose -f dev/docker-compose.yaml up --build`
+
+### Setup
+
```
Might be worth adding a step to make sure that we are ready for setup? For instance, making sure that opensearch is running/working?

nlisgo replied:
Might be worth adding a step to make sure that we are ready for setup? For instance, making sure that opensearch is running/working?

nlisgo reviewed (approved):
Will create a followup ticket to address the issues that are causing `project_tests.sh` not to work.

We should also consider adding kibana as this is a powerful elasticsearch client.
